src/HotChocolate/Core/src/Types/Configuration/TypeInitializer.cs

```diff
@@ -191,33 +192,58 @@ private void MergeTypeExtensions(DiscoveredTypes discoveredTypes)
                     .Where(t => t.IsNamedType)
                     .ToList();
 
-                foreach (RegisteredType extension in extensions)
+                foreach (IGrouping<NameString, RegisteredType> group in
+                    extensions.GroupBy(t => t.Type.Name))
                 {
                     RegisteredType type = types.FirstOrDefault(t =>
-                        t.Type.Name.Equals(extension.Type.Name));
+                        t.Type.Name.Equals(group.Key));
 
-                    if (type != null
-                        && extension.Type is INamedTypeExtensionMerger m
-                        && type.Type is INamedType n)
+                    if (type != null && type.Type is INamedType targetType)
                     {
-                        // merge
-                        CompletionContext context = _completionContext[extension];
-                        context.Status = TypeStatus.Named;
-                        MergeTypeExtension(context, m, n);
-
-                        // update dependencies
-                        context = _completionContext[type];
-                        type = type.AddDependencies(extension.Dependencies);
```

Comment:
The type was updated here but consequently the old type version was retrieved for each extension and overrode the updated type.